src/app/loadout-builder/process/useProcess.ts

```diff
@@ -282,22 +291,50 @@ function groupItems(
     }
   }
 
+  // First, map the DimItems to ProcessItems so that we can consider all things relevant to Loadout Optimizer.
+  const mappedItems: MappedItem[] = items.map((dimItem) => ({
+    dimItem,
+    processItem: mapDimItemToProcessItem({ dimItem, armorEnergyRules, modsForSlot }),
+  }));
+
   // Group by any restrictions so that items within the group can be freely exchanged subject
-  // to the user's mod choices. This means grouping by mod slot tags, mod energy capacity and type
+  // to the user's mod choices. This means grouping by mod slot tags, mod energy type
   // and exoticness. The groups for mod energies and tags are based on the mods
   // requested - for example if we only request solar mods, the groups should be
   // "solar", and "other", not "solar", "arc", "void", "stasis". If we only request
   // a vault of glass mod, we don't make a group that includes deep stone crypt mods.
-  const modGroupingFn = (item: DimItem) => {
+  const firstPassGroupingFn = ({
+    hash,
+    isExotic,
+    power,
+    compatibleModSeasons,
+    energy,
+    id: _id,
+    name: _name,
+    stats: _stats,
+    ...rest
+  }: ProcessItem) => {
+    // `id` is not relevant for grouping -- grouping is meant to consolidate different item instances.
+    // `name` is not relevant for grouping -- same reason as above.
+    // `stats` are handled in a second pass.
+    // `energy.val` is the same for all items, `energy.capacity` is handled in a second pass.
+
+    // This line fails to compile when more stuff is added to ProcessItem.
+    // When it does, decide whether the new property must be included in the grouping key
+    // or justify ignoring it with a comment above.
+    const exhaustiveCheck: Required<typeof rest> = {};
+    exhaustiveCheck[0]; // Really just here to silence an unused warning
+
     // Ensure exotics always form a distinct group
-    let groupId = item.isExotic ? `${item.hash}` : '';
+    let groupId = isExotic ? `${hash}` : '';
 
-    if (requiredActivityModSlots.size) {
-      const socketTags = getInterestingSocketMetadatas(item) ?? [];
-      for (const socketTag of socketTags) {
+    if (requiredActivityModSlots.size && compatibleModSeasons) {
+      for (const season of compatibleModSeasons) {
         // Only add to the grouping key if the socket matches what we need
-        if (requiredActivityModSlots.has(socketTag.slotTag)) {
-          groupId += socketTag.slotTag;
+        // FIXME: Why is it safe to ignore combat style mods? Do CWL/Warmind Cell mod slots
+        // that can't slot elemental well mods still exist?
+        if (requiredActivityModSlots.has(season)) {
+          groupId += season;
```

Comment:
I was going to write a comment here explaining why this is iffy and then I realized blue-quality items without a Combat Style mod slot exist so this needs fixing.